text/3730-nit-lint-level.md

```diff
@@ -0,0 +1,467 @@
+- Feature Name: `nit-lint-level`
+- Start Date: 2024-11-15
+- RFC PR: [rust-lang/rfcs#3730](https://github.com/rust-lang/rfcs/pull/3730)
+- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)
+- [Internals](https://internals.rust-lang.org/t/forbid-deny-warn-allow-and-notice/19986)
+
+# Summary
+[summary]: #summary
+
+Add a new visible lint level below `warn` to allow linters to act like a pair-programmer / code-reviewer where feedback is evaluated on a case-by-case basis.
+- `cargo` does not display these lints by default, requiring an opt-in
+- This RFC assumes LSPs/IDEs will opt-in
+- CIs could opt-in and open issues in the code review to show nits introduced in the current PR but not block merge on them
```

Comment:
Hm, I think I got a handle on one of my underlying concerns with this RFC: the _problem_ is one faced by a lot of different people regardless of the tooling they use, but the solution proposed is squarely in the realm of third-party IDE/CI support. Which isn't necessarily a non-starter, but I do think this problem _can_ be solved "in universe" in a way that benefits everyone.

IDEs are an optional thing one may use when working with Rust. So is `cargo`, admittedly, but `cargo` is much "less" optional and typically people replace cargo with an equivalent tool, unlike IDEs, which can be replaced with plain editors.


CI is also optional but since by and large the primary problem being solved here involves "what shows up locally vs what shows up in CI[^1]" it's not really a problem in a non-CI world.

However, choice of CI systems _is_ optional. Not every CI system will end up with this type of integrated tool, and the ones who do may not have one flexible enough to support folks' needs. People do a _lot_ of fancy stuff in their Rust CI and I wouldn't expect a "Run lints but only show nits on the CI" to tool to actually be usable by most users initially. As a datapoint, ICU4X has had the "show lints on PRs" CI task enabled and disabled a bunch of times for various reasons.


 [^1]: n.b: "shows up" here is about what the user actually notices: I'd consider `warn` lints as "showing up" locally, but _not_ "showing up" in CI, since people seldom look at the logs of successful CI runs.

Reply:
This RFC gives people the building blocks to make that policy.  If they don't implement it, they still have the ability to show these lints on the command-line.  Granted, that would be a firehose as pointed out in the RFC.  The community can explore and iterate on tools to help with this.

In the end, if the user does nothing extra, `nit` will be like `allow` and those users will have lost nothing from this.

Reply:
Okay, but I don't think "don't make the experience worse for other people" ought to be the bar for an RFC. This RFC lists problems that I have, that many people have, a its motivation. If the RFC does not solve those problems for us, I think that that's definitely an issue! It means the RFC is not fully solving the problems it talks about; and I don't think that's good for an RFC.

"Build your own CI tooling around this" is an acceptable conclusion for these folks. But "build your own CI tooling around this" was already a solution in the current world; this is just making that a _little_ easier.

Reply:
> Okay, but I don't think "don't make the experience worse for other people" ought to be the bar for an RFC. 

If this isn't solving a problem for some users, how is it making the problem worse for them?  The only thing I can think of is that linter teams decide to demote `warnings` into `nits` and they no longer show up.  Except there is already a common expectation of needing to do a lot of customization, so is that much of a change?

> If the RFC does not solve those problems for us, I think that that's definitely an issue! It means the RFC is not fully solving the problems it talks about; and I don't think that's good for an RFC.

Not every solution can solve a problem for every user.  Sometimes we intentionally don't try to solve problems for every user.

This is somewhat talked about in the other thread, like with https://github.com/rust-lang/rfcs/pull/3730#discussion_r1854365541

>  But "build your own CI tooling around this" was already a solution in the current world; this is just making that a little easier.

This is at least partially answered in the RFC.  To do this today, you need to use RUSTFLAGS.  If you just set the RUSTFLAGS in CI, you then can't reproduce it locally.

What you'd likely need to do is
- Maintain a `config.toml`, separate from your `[lints]`, for this
- Point rust-analyzer to this file through `--config`
  - Set up a separate `CARGO_TARGET_DIR` to avoid cache thrashing until rust-lang/cargo#14830 (if this third attempt even pans out)
- Point CI at this file through `--config`
- Document for people to know about this file and how to use it with `--config`

Even then to view the lints today you will get cache thrashing.  If rust-lang/cargo#14830 ends up working, you won't get cache thrashing but you won't get any cache reuse between these nit runs and regular runs.

CI will either need a separate run without the nits `config.toml` to run with `-Dwarnings` (requiring a full rebuild) or it will have to be the first thing your config does before showing turning all of the other allows into warnings.  With the latter, the user will then have r-a turning all standard warnings into errors which at minimum will block for anything built after.  if you then use `--keep-going`, it will only block dependents.  Maybe instead you have two separate `config.toml` files, one that has `-Dwarnings` and one that doesn't, requiring you to track two the nits in two places.  The r-a user then has the problem that there is no way to distinguish nits from regular warnings.

That is a complicated path requiring knowledge of various advanced features and likely not discovering some of the problems i called out until they are hit.  The burden for adopting this and maintaining it severely limits the likelihood someone would bother to adopt it.

So no, this is not about making it a *little* easier.  This is about smoothing the path so people are even willing to do it in the first place.

Reply:
Solving a problem for some but not all users often uses a lot of time and energy, which are usually both stretched quite thin within the Rust project.

A solution that suits some people can make a solution for the others become farther off because people don't always have the energy to do a thing, 'finally get it done', and then turn around and tackle almost the same problem all over again.

Reply:
> If this isn't solving a problem for some users, how is it making the problem worse for them? T

You misunderstand, I'm not saying it makes the problem worse for them, I just think that "doesn't make it worse for people" isn't the right bar for an RFC that's trying to solve a problem like this. You're free to disagree with that.


There is one way in which I'm worried this RFC makes things worse: people have wanted a `nit`-style "less than `warn`" level for _different_ reasons[^1] and I'm somewhat worried that having `nit` be used for this purpose would hinder the design of such a thing in the future. A design that explicitly declares that `CARGO_BUILD_NITS=deny` is never going to be a thing would definitely be a problem for a feature that lets you tweak lint UX for some lints but not others.


 [^1]: Specifically, having a level that indicates "different, less noisy UX", for example collapsing all lints down into a small report, or perhaps making them oneliners. Basically changing how they get reported. Which is kind of what's going on with the r-a and CI proposals here, but there are a lot more different ways the UX can be tweaked around.


> Not every solution can solve a problem for every user. Sometimes we intentionally don't try to solve problems for every user.
>
> This is somewhat talked about in the other thread, like with https://github.com/rust-lang/rfcs/pull/3730#discussion_r1854365541


Yes, but I'm not convinced this solves the problem for enough of the people who have it. As mentioned before there's a huge diversity of attitudes towards linting and I'm skeptical this is sufficient for them.

And to be clear, I'm not talking about other lint modality problems like "deny only when releasing" or "deny only when testing" as I was elsewhere, I'm talking about the problems listed in the motivation here.

Furthermore as @Lokathor mentioned, partial solutions do often legitimately hinder full solutions in the future.


> To do this today, you need to use RUSTFLAGS. 

No, that's not what I'm talking about here. This is in the context of PR-integrated CI systems (like "show errors on PRs", e.g. [clippy-action](https://github.com/giraffate/clippy-action))

This RFC proposes a feature that is useful if IDEs pick it up (and then, that's only useful to people using IDEs!), or if PR-integrated CI tooling systems pick it up. The CI thing isn't simply a use of `RUSTFLAGS` because for this to be surfaced to the user (people don't read non-failing CI logs) it needs to do the thing where the warnings are slurped up and tagged on the PR diff using the GitHub API (or whatever).

My current experience with that type of PR-integrated CI tooling is that it tends to work for simple things and often ends up needing much more configurability than is available when a codebase gets more complex. In which case you're kind of required to roll your own. Overall I don't see people doing that since it's a lot of work, instead when this happens folks just don't use the PR-integrated CI tooling, after all it's mostly just a QoL improvement: you can always figure out your lints from failing CI or a local run.

Thing is, if you're rolling your own anyway, then you can also fix the problems motivating this RFC by making the PR-integrated CI tool have a config for lints like this. That's relatively minor in the scheme of things when it comes to the amount of work involved in building one of these.

That's what I'm talking about when I say "it makes it a little easier". This RFC is predicated on PR-integrated CI tooling being built that consumes this, but you can do this _today_ in [clippy-action](https://github.com/giraffate/clippy-action)  using its flag settings. This RFC doesn't really enable anything _new_ for these systems, just a better convention that nicely integrates with `[lints]`. Which is valuable! It's far better for this to work with `[lints]` than some `clippy-action.toml` or `clippy_flags: -Dfoo -Abar` or whatever since it's great for `[lints]` to be a one-stop shop, and I really don't want to undersell the benefit of that!

But! If a PR-integrated CI system doesn't work for me, I have to roll my own _anyway_, at which point this better convention doesn't buy me much in the scheme of things. That's what i mean by "it makes it a little easier", I'm talking about people for whom the existing PR-integrated CI systems don't work well. I don't think that's a small set of people at all, I've seen relatively low uptake of these systems, partially because of permissions issues, partly because of more bespoke build setups, etc.

It makes it a lot easier for people for whom this style of integrated CI tooling works well out of the box. I'm not convinced that's a lot of people.

Reply:
> Sometimes we intentionally don't try to solve problems for every user.



Specifically about this: my point here is to highlight that users who care about replicable CI, users who don't use IDEs, and users for whom PR-integrated CI doesn't work are not covered by this RFC, and if we are being _intentionally_ excluded from the solutions here, that is fine, but that _intention_ needs to be there, and when the lang team makes a call on this they should be aware, and attempt to understand what percentage of the target audience of this RFC's motivation is _intentionally_ excluded from the solution.


I don't want it to fly under the radar that these groups are not going to benefit from this RFC. You're absolutely right that we sometimes intentionally don't solve problems for every user. However, we actually do have to be intentional about it.